packages/playwright-core/src/server/chromium/crNetworkManager.ts

```diff
@@ -40,15 +43,22 @@ export class CRNetworkManager {
   private _eventListeners: RegisteredListener[];
   private _responseExtraInfoTracker = new ResponseExtraInfoTracker();
 
-  constructor(client: CRSession, page: Page, parentManager: CRNetworkManager | null) {
+  constructor(client: CRSession, page: Page | null, serviceWorker: CRServiceWorker | null, parentManager: CRNetworkManager | null) {
     this._client = client;
     this._page = page;
+    this._serviceWorker = serviceWorker;
     this._parentManager = parentManager;
     this._eventListeners = this.instrumentNetworkEvents(client);
   }
 
+  setParentManager(parent: CRNetworkManager) {
+    if (this._parentManager && this._parentManager !== parent)
+      throw new Error('A different parent manager was already set. Cannot overwrite.');
+    this._parentManager = parent;
```

Comment:
This seems like a nice place to call `_maybeAdoptServiceWorkerMainRequest` and call `_onRequestWillBeSent` with that event.